rust: Add support of MacVlan and MacVtap

cathay4t opened the pull request:
Integration test cases enabled.

thom311 commented on rust/src/libnm_dbus/device.rs:
```diff
@@ -481,18 +482,46 @@ fn nm_dev_state_reason_get(
+        is_mac_vtap: false,
+    };
+    if dev.iface_type == "macvlan" {
+        dev.is_mac_vtap = nm_dev_is_mac_vtap_get(dbus_conn, obj_path)?;
```
hm. This approach with zbus' proxies seems expensive.

No question that blocking calls result in clean, readable code. On the surface at last. It's not very obvious/readable that such functions contain blocking IO. Anyway, it seems expensive and subtly wrong (like all blocking D-Bus calls are fundamentally flawed because they don't preserve the ordering of events on D-Bus -- https://smcv.pseudorandom.co.uk/2008/11/nonblocking/ ).

This function `nm_dev_from_obj_path()` calls three functions, that each create (and initialize) a D-Bus proxy object. That means, the function requires at least 3 blocking D-Bus method calls (that is, round trips). Possibly even more, because initializing a D-Bus proxy object might require more than D-Bus calls (like first an AddMatch or GetNameOwner, dunno)...

The three creations could be avoided in part, by creating the proxy object earlier and passing it to the getter functions. Or even let the caller provide the proxy object already (as the caller might have another use of the proxy object to further). Basically, cache the proxy object and reuse the things.

thom311 replied:
https://docs.rs/zbus/1.9.2/zbus/struct.Proxy.html#current-limitations

ah. OK, so properties are not cached and the name owner is not tracked. I guess that means every `get_property()` call translates to a `Get()` D-Bus call...

thom311 replied:
ok, rethinking...

if the zbus proxy does no caching, then these `get_property()` calls are plain `Get` D-Bus calls (at which point it's not clear what the benefit of the proxy object is, you could just as well make a zbus `call()`). 

Anyway. You need three properties, and you either need 3 `Get()` calls, or one `GetAll()` / `GetManagedObjects()` call. With `GetAll()` (and especially `GetManagedObjects()`) you fetch more than you need, but only have one round-trip. It's not clear which one is better... you anyway will always require at least one roundtrip.

Btw, libnm only makes one `GetManagedObjects()` call at the beginning (to get the initial state). Afterwards, it only relies on the PropertiesChanged signals to get updates. That it needs to maintain a cache of everything, that you might use in the future. The downside is that much of the data you might never actually use and the effort is wasted.

But making 3 `Get()` requests one after the other, is what causes this to be 3 times slower than if you could fire them off in parallel. If you think further and are initializing many `NmDevice`, then each of them requires 3 (or one 1) round trip, and the real gain would be, if you could fire off the initialization of them all in parallel. That would require some sort of `async` implementation...

cathay4t replied:
Currently we see no problem on this sync code and we suffered enough in python for the async way. For the getall suggestions, I will use it if someday we noticed a big impact. For the direct call of zbus `call`, again no issue no change.

ffmancera commented on rust/src/lib/iface.rs:
```diff
@@ -440,26 +470,21 @@ impl Interface {
     pub(crate) fn pre_edit_cleanup(&mut self) -> Result<(), NmstateError> {
-        self.base_iface_mut().pre_edit_cleanup()
+        self.base_iface_mut().pre_edit_cleanup()?;
+        if let Interface::MacVlan(iface) = self {
+            iface.validate()?;
+        } else if let Interface::MacVtap(iface) = self {
+            iface.validate()?;
+        }
+        Ok(())
```
Why is this required? For the other interfaces we are not validating during the `pre_edit_cleanup`.. I think this should be on `validate()` instead on `src/lib/iface.rs`.

cathay4t replied:
Moved to `Interface::validate()`

ffmancera commented on rust/src/lib/nm/show.rs:
```diff
@@ -155,52 +168,57 @@ pub(crate) fn nm_retrieve() -> Result<NetworkState, NmstateError> {
     }
 }
 
-fn nm_conn_to_base_iface(nm_conn: &NmConnection) -> Option<BaseInterface> {
+fn nm_conn_to_base_iface(
+    nm_dev: &NmDevice,
+    nm_conn: &NmConnection,
+) -> Option<BaseInterface> {
     if let Some(iface_name) = nm_conn.iface_name() {
-        if let Some(iface_type) = nm_conn.iface_type() {
-            let ipv4 = nm_conn.ipv4.as_ref().map(|nm_ipv4_setting| {
-                nm_ip_setting_to_nmstate4(nm_ipv4_setting)
-            });
-            let ipv6 = nm_conn.ipv6.as_ref().map(|nm_ipv6_setting| {
-                nm_ip_setting_to_nmstate6(nm_ipv6_setting)
-            });
+        let ipv4 = nm_conn.ipv4.as_ref().map(nm_ip_setting_to_nmstate4);
+        let ipv6 = nm_conn.ipv6.as_ref().map(nm_ip_setting_to_nmstate6);
 
-            let mut base_iface = BaseInterface::new();
-            base_iface.name = iface_name.to_string();
-            base_iface.prop_list =
-                vec!["name", "state", "iface_type", "ipv4", "ipv6"];
-            base_iface.state = InterfaceState::Up;
-            base_iface.iface_type = nm_iface_type_to_nmstate(iface_type);
-            base_iface.ipv4 = ipv4;
-            base_iface.ipv6 = ipv6;
-            base_iface.controller = nm_conn.controller().map(|c| c.to_string());
-            return Some(base_iface);
-        }
+        let mut base_iface = BaseInterface::new();
+        base_iface.name = iface_name.to_string();
+        base_iface.prop_list =
+            vec!["name", "state", "iface_type", "ipv4", "ipv6"];
+        base_iface.state = InterfaceState::Up;
+        base_iface.iface_type = nm_dev_iface_type_to_nmstate(nm_dev);
+        base_iface.ipv4 = ipv4;
+        base_iface.ipv6 = ipv6;
+        base_iface.controller = nm_conn.controller().map(|c| c.to_string());
+        return Some(base_iface);
```
I agree with these changes but I do not think they are related to this commit. Could you move this to its own commit. You can keep it on this PR IMO.

cathay4t replied:
actually, they are related to this PR. This is the new way on checking interface type, previously we just use `nm_iface_type`, but for MacVtap, we need `nm_dev` to detemine whether is it MacVlan or MacVtap.

Anyway, I will split it out.

cathay4t replied:
Patch isolated.

ffmancera reviewed (changes requested):
Thanks! In general it looks good but I made two comments. Let me know what do you think!

cathay4t committed:
rust: Add support of MacVlan and MacVtap

Integration test cases enabled.

Signed-off-by: Gris Ge <fge@redhat.com>

cathay4t committed:
rust nm: Use NmDevice to determine the interface type

In order to support MacVlan and MacVtap in the future, we need to
pass NM.Device when determine the interface type.

Signed-off-by: Gris Ge <fge@redhat.com>

ffmancera reviewed (approved):
LGTM! Thanks!
